src/cbs.c

```diff
@@ -7,153 +7,342 @@
 #include <string.h>
 #include "azure_c_shared_utility/optimize_size.h"
 #include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/singlylinkedlist.h"
 #include "azure_uamqp_c/cbs.h"
 #include "azure_uamqp_c/amqp_management.h"
 #include "azure_uamqp_c/session.h"
 
+typedef enum CBS_STATE_TAG
+{
+    CBS_STATE_CLOSED,
+    CBS_STATE_OPENING,
+    CBS_STATE_OPEN,
+    CBS_STATE_ERROR
+} CBS_STATE;
+
+typedef struct CBS_OPERATION_TAG
+{
+    ON_CBS_OPERATION_COMPLETE on_cbs_operation_complete;
+    void* on_cbs_operation_complete_context;
+    SINGLYLINKEDLIST_HANDLE pending_operations;
+} CBS_OPERATION;
+
 typedef struct CBS_INSTANCE_TAG
 {
     AMQP_MANAGEMENT_HANDLE amqp_management;
+    CBS_STATE cbs_state;
+    ON_CBS_OPEN_COMPLETE on_cbs_open_complete;
+    void* on_cbs_open_complete_context;
+    ON_CBS_ERROR on_cbs_error;
+    void* on_cbs_error_context;
+    SINGLYLINKEDLIST_HANDLE pending_operations;
 } CBS_INSTANCE;
 
 static int add_string_key_value_pair_to_map(AMQP_VALUE map, const char* key, const char* value)
 {
     int result;
 
     AMQP_VALUE key_value = amqpvalue_create_string(key);
-    if (key == NULL)
+    if (key_value == NULL)
     {
+        /* Codes_SRS_CBS_01_072: [ If constructing the message fails, `cbs_put_token_async` shall fail and return a non-zero value. ]*/
+        LogError("Failed creating value for property key %s", key);
         result = __FAILURE__;
     }
     else
     {
         AMQP_VALUE value_value = amqpvalue_create_string(value);
```

Comment:
Yeah. I guess. I'm going to leave it as such. I did consider changing it, but I got lazy between the CR and the IDE.

---
In reply to: [107037103](https://github.com/Azure/azure-uamqp-c/pull/56#discussion_r107037103) [](ancestors = 107037103)